docs/eBpfExtensions.md

```diff
@@ -1,42 +1,207 @@
+
+```
+/**
+ *  @brief This is the only function in the eBPF hook NPI client dispatch table.
+ */
+typedef ebpf_result_t (*ebpf_invoke_program_function_t)(
+    _In_ const void* client_binding_context, _In_ const void* context, _Out_ uint32_t* result);
+
+```
+The function pointer can be obtained from the client dispatch table as follows:
+```
+invoke_program = (ebpf_invoke_program_function_t)client_dispatch_table->function[0];
+```
+When an extension invokes this function pointer, then the call flows through the eBPF Execution Context and eventually invokes the eBPF program.
+When invoking an eBPF program, the extension must supply the client binding context it obtained from the hook NPI client as the `client_binding_context` parameter. For the second parameter `context`, it must pass the program type specific context data structure. Note that the program information NPI provider supplies information the context descriptor using `ebpf_context_descriptor_t` type to the eBPF verifier and JIT-compiler via the NPI client hosted by Execution Context. The `result` output parameter holds the return value from the eBPF program post execution.
+
+### 3.3 Authoring Helper Functions
+An extension can provide implementation of helper functions that can be invoked by the eBPF programs. The helper functions can be of two types:
+1. Program-Type specific: These helper functions can only be invoked by eBPF programs of a given program type. Usually, an extension may provide implementations for hooks of certain program types and provide helper functions that are associated with those helper functions. The program information NPI provider must then provide the prototypes and addresses for those functions. For these type of helpers, the helper function Id must be greater that 65535 (0xFFFF) for program type specific helper functions.
+2. General: The general helper functions can be invoked by eBPF programs of all types. Examples of this type of helper functions are the eBPF Map helper functions. These helper functions are implemented by the eBPF Execution Context itself. However, if a program type so chooses, it may provide implementations for general helper functions. For that the extension would have to provide another program information NPI provider, which *does not* provide any program context descriptor. Instead, it only supplies the prototypes and addresses of the general helper functions. The NPI ID of this module defined as:
+```
+GUID ebpf_general_helper_function_interface_id = {/* 8d2a1d3f-9ce6-473d-b48e-17aa5c5581fe */
+                                                  0x8d2a1d3f,
+                                                  0x9ce6,
+                                                  0x473d,
+                                                  {0xb4, 0x8e, 0x17, 0xaa, 0x5c, 0x55, 0x81, 0xfe}};
+```
+The helper function Id for a general helper function must be in the range 0 - 65535 and must be globally unique.
```

Comment:
```suggestion
The helper function ID for a general helper function must be in the range 0 - 65535 and must be globally unique.
```
What process should an extension developer use to get a unique number that won't be used by any other such extension?